### PR TITLE
Add io.github.theforceengine.tfe

### PR DESCRIPTION
**Pull request by fpiesche**
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [x] My pull request follows the instructions at [App Submission][submission].
- [x] I have [built][build] and tested the submission locally.
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [x] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:** https://github.com/luciusDXL/TheForceEngine/issues/268#issuecomment-1904723735
- [x] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*
  - `game_data_detection.patch` is a minimum-effort implementation to make the detection work on Linux (which it doesn't at all by default). I'm not a skilled C++ programmer so I was struggling to implement the feature properly rather than just replacing the Windows detection entirely but am planning to experiment more and submit this upstream if I can get somewhere.
  - `user_data.patch` - The lack of user/application data separation is an artifact of TFE's primary focus being Windows so far. Again this is a minimum-effort fix that makes things work in the context of a sandboxed Flatpak distribution but I'm not a good enough C++ programmer to confidently implement this to work in other potential setups.
  - The AppStream metadata file has been submitted upstream and is awaiting merge.

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

### CODE REVIEW

File: io.github.theforceengine.tfe.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+finish-args:
+  # hardware 3D
+  - --device=dri
+  # X11 + XShm access
+  - --share=ipc
+  - --socket=x11
+  # Audio
+  - --socket=pulseaudio
+  # Access to the default Steam Dark Forces data directory
+  - --filesystem=~/.local/share/Steam/steamapps/common/Dark Forces/Game/
```
Comment (hfiguiere) [inline io.github.theforceengine.tfe.yml:16]: how does this work of one uses the flatpak version of steam?

Also does it work the the space in the filename?

File: io.github.theforceengine.tfe.yml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        commands:
+          - # Check for game data
+          - if [[ -z "$TFE_SKIP_DATACHECK" ]] && [[ -z $(find "$XDG_DATA_HOME" -iname dark.gob) ]] && [[ ! -d "$HOME/.local/share/Steam/steamapps/common/Dark Forces/Game/" ]]; then
+          -     zenity --error --text "<b>Could not find Dark Forces game data</b>\n\n
+                    Please either install Dark Forces via Steam to your library in <tt><b>$HOME/.local/share/Steam/</b></tt>
+                    or copy the game data to <tt><b>$XDG_DATA_HOME/</b></tt>."
+                    --ok-label "Close" --width=400
+          - fi
+          - export TFE_DATA_HOME=/app/share/TheForceEngine
+          - theforceengine
```
Comment (hfiguiere) [inline io.github.theforceengine.tfe.yml:52]:
```suggestion
          - exec theforceengine
```

File: io.github.theforceengine.tfe.yml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    builddir: true
+    build-commands:
+      # Install launcher script
+      - install -Dm 744 /run/build/tfe/io.github.theforceengine.tfe.sh -t /app/bin
+      # Edit and install desktop launcher file
+      - desktop-file-edit --set-key=Icon --set-value=io.github.theforceengine.tfe /run/build/tfe/TheForceEngine/TheForceEngine.desktop
+      - desktop-file-edit --set-key=Exec --set-value=io.github.theforceengine.tfe.sh /run/build/tfe/TheForceEngine/TheForceEngine.desktop
+      - install -Dm 644 /run/build/tfe/TheForceEngine/TheForceEngine.desktop /app/share/applications/io.github.theforceengine.tfe.desktop
+      # Install icon
+      - install -Dm 644 /run/build/tfe/TheForceEngine/TheForceEngine.png /app/share/icons/hicolor/256x256/apps/io.github.theforceengine.tfe.png
```
Comment (hfiguiere) [inline io.github.theforceengine.tfe.yml:65]:
```suggestion
      - install -Dm 644 ../TheForceEngine/TheForceEngine.png /app/share/icons/hicolor/256x256/apps/io.github.theforceengine.tfe.png
```

File: io.github.theforceengine.tfe.yml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      # Install locally-included appstream metadata file until a TFE release is made that includes it
+      # https://github.com/luciusDXL/TheForceEngine/pull/375
+      - install -Dm 644 /run/build/tfe/io.github.theforceengine.tfe.appdata.xml /app/share/metainfo/io.github.theforceengine.tfe.appdata.xml
```
Comment (hfiguiere) [inline io.github.theforceengine.tfe.yml:68]:
```suggestion
      - install -Dm 644 ../io.github.theforceengine.tfe.appdata.xml /app/share/metainfo/io.github.theforceengine.tfe.appdata.xml
```

File: io.github.theforceengine.tfe.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    buildsystem: cmake-ninja
+    config-opts:
+      - -DDISABLE_SYSMIDI=ON
+    builddir: true
+    build-commands:
+      # Install launcher script
+      - install -Dm 744 /run/build/tfe/io.github.theforceengine.tfe.sh -t /app/bin
+      # Edit and install desktop launcher file
+      - desktop-file-edit --set-key=Icon --set-value=io.github.theforceengine.tfe /run/build/tfe/TheForceEngine/TheForceEngine.desktop
+      - desktop-file-edit --set-key=Exec --set-value=io.github.theforceengine.tfe.sh /run/build/tfe/TheForceEngine/TheForceEngine.desktop
```
Comment (hfiguiere) [inline io.github.theforceengine.tfe.yml:62]:
```suggestion
      - desktop-file-edit --set-key=Exec --set-value=io.github.theforceengine.tfe.sh ../TheForceEngine/TheForceEngine.desktop
```

File: io.github.theforceengine.tfe.yml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+          - theforceengine
+    buildsystem: cmake-ninja
+    config-opts:
+      - -DDISABLE_SYSMIDI=ON
+    builddir: true
+    build-commands:
+      # Install launcher script
+      - install -Dm 744 /run/build/tfe/io.github.theforceengine.tfe.sh -t /app/bin
+      # Edit and install desktop launcher file
+      - desktop-file-edit --set-key=Icon --set-value=io.github.theforceengine.tfe /run/build/tfe/TheForceEngine/TheForceEngine.desktop
```
Comment (hfiguiere) [inline io.github.theforceengine.tfe.yml:61]:
```suggestion
      - desktop-file-edit --set-icon=io.github.theforceengine.tfe ../TheForceEngine/TheForceEngine.desktop
```

File: io.github.theforceengine.tfe.yml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      - desktop-file-edit --set-key=Exec --set-value=io.github.theforceengine.tfe.sh /run/build/tfe/TheForceEngine/TheForceEngine.desktop
+      - install -Dm 644 /run/build/tfe/TheForceEngine/TheForceEngine.desktop /app/share/applications/io.github.theforceengine.tfe.desktop
```
Comment (hfiguiere) [inline io.github.theforceengine.tfe.yml:63]:
```suggestion
      - install -Dm 644 ../TheForceEngine/TheForceEngine.desktop /app/share/applications/io.github.theforceengine.tfe.desktop
```

**Comment by hfiguiere** [comment]
bot, build io.github.theforceengine.tfe

File: io.github.theforceengine.tfe.appdata.xml [hunk 8]
```diff
@@ -0,0 +1,53 @@
+ <?xml version="1.0" encoding="UTF-8"?>
+<application>
+  <id type="desktop">io.github.theforceengine.tfe</id>
+  <name>The Force Engine</name>
+  <developer_name>The Force Engine Contributors</developer_name>
+  <summary>A modern cross-platform port of Star Wars: Dark Forces with mod support</summary>
```
Comment (bbhtt) [inline io.github.theforceengine.tfe.appdata.xml:6]:
```suggestion
  <summary>Modern cross-platform port of Star Wars: Dark Forces with mod support</summary>
```

https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/quality-guidelines#dont-start-with-an-article

File: io.github.theforceengine.tfe.appdata.xml
Comment (bbhtt) [inline io.github.theforceengine.tfe.appdata.xml:1]: This file should be upstream.

Comment (fpiesche) [reply io.github.theforceengine.tfe.appdata.xml:1]: I have submitted an upstream PR to include the file there which is awaiting a merge.  As this build is currently aiming at tagged releases I've included it locally so the current release version can be distributed as a Flatpak; once the PR upstream is merged I'm planning to remove the local file so once the next release gets tagged the upstream file will be used instead.

File: io.github.theforceengine.tfe.yml [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    builddir: true
+    build-commands:
+      # Install launcher script
+      - install -Dm 744 ../io.github.theforceengine.tfe.sh -t /app/bin
+      # Edit and install desktop launcher file
+      - desktop-file-edit --set-icon=io.github.theforceengine.tfe ../TheForceEngine/TheForceEngine.desktop
+      - desktop-file-edit --set-key=Exec --set-value=io.github.theforceengine.tfe.sh ../TheForceEngine/TheForceEngine.desktop
+      - install -Dm 644 ../TheForceEngine/TheForceEngine.desktop /app/share/applications/io.github.theforceengine.tfe.desktop
+      # Install icon
+      - install -Dm 644 ../TheForceEngine/TheForceEngine.png /app/share/icons/hicolor/256x256/apps/io.github.theforceengine.tfe.png
```
Comment (hfiguiere) [inline io.github.theforceengine.tfe.yml:65]: This is unnecessary:
```
-- Installing: /app/share/icons/hicolor/256x256/apps/TheForceEngine.png
```

So is line 63. You can use `rename-icon` and `rename-desktop-file` to rename them. `rename-icon` also make line 61 unecessary.

File: io.github.theforceengine.tfe.yml (hunk 1 above)
Comment (fpiesche) [reply io.github.theforceengine.tfe.yml:16]: For a regular, non-Flatpak/Snap installation of Steam, this works fine even with the space in the path - I've now extended the permissions and launcher wrap script to also check the default library path for the Flatpak version of Steam for the game data.

There are a number of hard to cover cases for the detection here - if e.g. the user has the game installed to another Steam library that isn't in the default location (say, on the SD card on a Steam Deck) this path will also be different and the auto-detection will fail.

Users can always copy the game data to `~/.var/app/io.github.theforceengine.tfe/data/` if the auto-detection does not work (and will in fact be instructed to do so by the launcher script); I just wanted to cover the "default" option.

File: io.github.theforceengine.tfe.yml [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    builddir: true
+    build-commands:
+      # Install launcher script
+      - install -Dm 744 ../io.github.theforceengine.tfe.sh -t /app/bin
+      # Edit and install desktop launcher file
+      - desktop-file-edit --set-key=Exec --set-value=io.github.theforceengine.tfe.sh ../TheForceEngine/TheForceEngine.desktop
+      # Install locally-included appstream metadata file until a TFE release is made that includes it
+      # https://github.com/luciusDXL/TheForceEngine/pull/375
+      - install -Dm 644 ../io.github.theforceengine.tfe.metainfo.xml /app/share/metainfo/io.github.theforceengine.tfe.metainfo.xml
+      # - install -Dm 644 /run/build/tfe/TheForceEngine/TheForceEngine.metainfo.xml /app/share/metainfo/io.github.theforceengine.tfe.metainfo.xml
```
Comment (hfiguiere) [inline io.github.theforceengine.tfe.yml:68]:
```suggestion
```

Comment (fpiesche) [reply io.github.theforceengine.tfe.yml:68]: Ah, does flatpak-builder automatically detect the metainfo files with that filename?

Comment (hfiguiere) [reply io.github.theforceengine.tfe.yml]: this line is a comment, you don't need it. there is no detection, it's installed by the previous line.

Comment (fpiesche) [reply io.github.theforceengine.tfe.yml]: Yeah, I realised that as soon as I wrote my reply :D the commented-out line is what it'll want to be once TFE merge the metainfo and tag a release but I'm sure I'll be able to remember that when the time comes.

Comment (hfiguiere) [reply io.github.theforceengine.tfe.yml]: it should be done by the build system. this is not specific to flatpak.

(`/app` is just the prefix)

Comment (fpiesche) [reply io.github.theforceengine.tfe.yml]: Will add that to their cmakelists in my PR, then!

**Comment by barthalion** [comment]
bot, build io.github.theforceengine.tfe

**Pull request closed without merging**